# Problems

## 1-1 Comparison of running times

> For each function $f(n)$ and time $t$ in the following table, determine the largest size $n$ of a problem that can be solved in time $t$, assuming that the algorithm to solve the problem takes $f(n)$ microseconds.
>
> |            | 1 second | 1 minute | 1 hour | 1 day | 1 month | 1 year | 1 century |
> |------------|----------|----------|--------|-------|---------|--------|-----------|
> | $\lg n$    |          |          |        |       |         |        |           |
> | $\sqrt{n}$ |          |          |        |       |         |        |           |
> | $n$        |          |          |        |       |         |        |           |
> | $n\lg n$   |          |          |        |       |         |        |           |
> | $n^2$      |          |          |        |       |         |        |           |
> | $n^3$      |          |          |        |       |         |        |           |
> | $2^n$      |          |          |        |       |         |        |           |
> | $n!$       |          |          |        |       |         |        |           |

We assume that a month has 30 days and a year has 365 days. Note that $1\text{ second}=10^6\text{ microseconds}$. So each time (column) can be rewritten in terms of microseconds.

Thus, for each function $f(n)$ and each time $t$, we need to find the largest $n$ for which $f(n)\leq t$. The first three rows can be solved analitically:

- $\lg n \leq t\iff n\leq 2^t$.
- $\sqrt{n} \leq t \iff n \leq t^2$.
- $n\leq t$ is already solved.

The other rows need to be solved computationally. We implement C code to complete the table.

In [1]:
#include <stdio.h>
#include <math.h>

unsigned long long int pow2( int n ) {
    if (n==0) {
        return 1;
    }
    return 2*pow2(n-1);
}

unsigned long long int Pow10( int n ) {
    if (n==0) {
        return 1;
    }
    return 10*Pow10(n-1);
}

unsigned long long fact( int n ) {
    if (n==1) {
        return 1;
    }
    return n*fact(n-1);
}

int main () {
    int i,j; // index
    unsigned long long times[7] = {Pow10(6), // 1 second = 10^6 usec
                                   Pow10(6)*60, // 1 minute = 6*10^7 usec
                                   Pow10(6)*60*60, // 1 hour = 36 * 10^8 usec
                                   Pow10(6)*60*60*24, // 1 day = 864 * 10^8 usec
                                   Pow10(6)*60*60*24*30, // 1 month = 2592 * 10^9 usec
                                   Pow10(6)*60*60*24*365, // 1 year = 31536 * 10^9 usec
                                   Pow10(6)*60*60*24*365*100 // 1 century = 31536 * 10^11 usec
                                   };
    unsigned long long t;
    char format_string[20];
    int col_widths[7] = {16,17,19,20,22,23,25};
    int first_col_width = 10;
    
    char initial_rows[5][8][30]={
            {"","1 second","1 minute","1 hour","1 day","1 month","1 year","1 century"},
            {"---","---","---","---","---","---","---","---"},
            {"$\\lg n$"},
            {"$\\sqrt{n}$"},
            {"$n$"},
        };
    
    for (i=1;i<8;i++) {
        sprintf(initial_rows[2][i],"$2^{%llu}$",times[i-1]);
        sprintf(initial_rows[3][i],"$\\sqrt{%llu}$",times[i-1]);
        sprintf(initial_rows[4][i],"$%llu$",times[i-1]);
    }
    for (i=0;i<5;i++) {
        printf("|");
        // Little trick to allow printing formatted string with parameters in a variable
        sprintf(format_string," %c%ds |",'%',first_col_width);
        printf(format_string,initial_rows[i][0]);
        for (j=1;j<8;j++) {
            sprintf(format_string," %c%ds |",'%',col_widths[j-1]);
            printf(format_string,initial_rows[i][j]);
        }
        printf("\n");
    }
    
    // For the next ones, the largest n for which f(n)<=t will be found with the bisection method.
    // We can use the interval [0,t], as the functions always satisfy f(t)>t>f(0)=1 or 0
    unsigned long long n_l,n_u,mi; //lower n, upper n, middle
    
    // Solve n lg n <=t
    
    sprintf(format_string,"| %c%ds |",'%',first_col_width);
    printf(format_string,"$n\\lg n$");
    for (i=0;i<7;i++) {
        t=  times[i];
        n_l = 0;
        n_u = t;
        while (n_l+1<n_u) {
            mi = (n_l+n_u)/2;
             if (mi * log2((double)mi) <= t) {
                n_l=mi;
            } else {
                n_u=mi;
            }
        }
        
        sprintf(format_string," %c%dllu |",'%',col_widths[i]);
        printf(format_string,n_l);
    }
    
    printf("\n");
        
    // Solve n^2 <=t
    
    sprintf(format_string,"| %c%ds |",'%',first_col_width);
    printf(format_string,"$n^2$");
    for (i=0;i<7;i++) {
        t=  times[i];
        n_l = 0;
        n_u = t;
        while (n_l+1<n_u) {
            mi = (n_l+n_u)/2;
             if (mi * mi <= t) {
                n_l=mi;
            } else {
                n_u=mi;
            }
        }
        
        sprintf(format_string," %c%dllu |",'%',col_widths[i]);
        printf(format_string,n_l);
    }
    
    printf("\n");
        
    // Solve n^3 <=t
    
    sprintf(format_string,"| %c%ds |",'%',first_col_width);
    printf(format_string,"$n^3$");
    for (i=0;i<7;i++) {
        t=  times[i];
        n_l = 0;
        n_u = t;
        while (n_l+1<n_u) {
            mi = (n_l+n_u)/2;
             if (mi * mi * mi<= t) {
                n_l=mi;
            } else {
                n_u=mi;
            }
        }
        
        sprintf(format_string," %c%dllu |",'%',col_widths[i]);
        printf(format_string,n_l);
    }
    
    printf("\n");
        
    // Solve 2^n <=t
    
    sprintf(format_string,"| %c%ds |",'%',first_col_width);
    printf(format_string,"$2^n$");
    for (i=0;i<7;i++) {
        t=  times[i];
        n_l = 0;
        /*
            Here, taking n_u = t would entail calculating 2^t, which is too large.
            We can choose a smaller upper bound: The largest t is
                t = 10^6*60*60*24*365*100
                  = (2^6 * 5^6) *(3 * 5 * 2^2) * (3 * 5 * 2^2) * (2^3 * 3) * (73 * 5) * (2^2 * 5^2)
                  = 2^15 * 3^3 * 5^11 * 73
                  < 2^15 * 3^3 * (5^3)^4 * 73
                  = 2^15 * 27 * (125)^4 * 73
                  < 2^15 * 32 * (128)^4 * 128
                  = 2^15 * 2^5 * (2^7)^4 * 2^7
                  = 2^55
            so taking n_u=55 is more than enough.
        */
        n_u = 55;
        while (n_l+1<n_u) {
            mi = (n_l+n_u)/2;
             if (pow2(mi)<= t) {
                n_l=mi;
            } else {
                n_u=mi;
            }
        }
        
        sprintf(format_string," %c%dllu |",'%',col_widths[i]);
        printf(format_string,n_l);
    }
    
    printf("\n");
    
    // Solve n! <=t
    
    sprintf(format_string,"| %c%ds |",'%',first_col_width);
    printf(format_string,"$n!$");
    
    /*
        In this case, as factorials grow extremely fast, we can simply do a linear search starting at 1.
        
        Compiling the previous row already lets us know that n<=51 (as we also know that 51!>2^51, clearly).
        
        We could also implement factorials implicity here for better performance, but the code below is more readable.
    */
    n_l=1;
    
    for (i=0;i<7;i++) {
        t = times[i];
        
        while (fact(n_l)<=t) n_l++;
        sprintf(format_string," %c%dllu |",'%',col_widths[i]);
        printf(format_string,n_l-1);
    }
    
    printf("\n");
    
    return 0;
}     

|            |         1 second |          1 minute |              1 hour |                1 day |                1 month |                  1 year |                 1 century |
|        --- |              --- |               --- |                 --- |                  --- |                    --- |                     --- |                       --- |


|    $\lg n$ |    $2^{1000000}$ |    $2^{60000000}$ |    $2^{3600000000}$ |    $2^{86400000000}$ |    $2^{2592000000000}$ |    $2^{31536000000000}$ |    $2^{3153600000000000}$ |
| $\sqrt{n}$ | $\sqrt{1000000}$ | $\sqrt{60000000}$ | $\sqrt{3600000000}$ | $\sqrt{86400000000}$ | $\sqrt{2592000000000}$ | $\sqrt{31536000000000}$ | $\sqrt{3153600000000000}$ |
|        $n$ |        $1000000$ |        $60000000$ |        $3600000000$ |        $86400000000$ |        $2592000000000$ |        $31536000000000$ |        $3153600000000000$ |
|   $n\lg n$ |            62746 |           2801417 |           133378058 |           2755147513 |            71870856404 |            797633893349 |            68610956750570 |
|      $n^2$ |             1000 |              7745 |               60000 |               293938 |                1609968 |                 5615692 |                  56156922 |
|      $n^3$ |              100 |               391 |                1532 |                 4420 |            

Thus, the completed table is

|            |         1 second |          1 minute |              1 hour |                1 day |                1 month |                  1 year |                 1 century |
|        --- |              --- |               --- |                 --- |                  --- |                    --- |                     --- |                       --- |
|    $\lg n$ |    $2^{1000000}$ |    $2^{60000000}$ |    $2^{3600000000}$ |    $2^{86400000000}$ |    $2^{2592000000000}$ |    $2^{31536000000000}$ |    $2^{3153600000000000}$ |
| $\sqrt{n}$ | $\sqrt{1000000}$ | $\sqrt{60000000}$ | $\sqrt{3600000000}$ | $\sqrt{86400000000}$ | $\sqrt{2592000000000}$ | $\sqrt{31536000000000}$ | $\sqrt{3153600000000000}$ |
|        $n$ |        $1000000$ |        $60000000$ |        $3600000000$ |        $86400000000$ |        $2592000000000$ |        $31536000000000$ |        $3153600000000000$ |
|   $n\lg n$ |            62746 |           2801417 |           133378058 |           2755147513 |            71870856404 |            797633893349 |            68610956750570 |
|      $n^2$ |             1000 |              7745 |               60000 |               293938 |                1609968 |                 5615692 |                  56156922 |
|      $n^3$ |              100 |               391 |                1532 |                 4420 |                  13736 |                   31593 |                    146645 |
|      $2^n$ |               19 |                25 |                  31 |                   36 |                     41 |                      44 |                        51 |
|       $n!$ |                9 |                11 |                  12 |                   13 |                     15 |                      16 |                        17 |